src/file-upload/internal.tsx

```diff
@@ -0,0 +1,181 @@
+import { InternalBaseComponentProps } from '../internal/hooks/use-base-component';
+
+import { FileOption } from './file-option';
+import { ButtonProps } from '../button/interfaces';
+import InternalSpaceBetween from '../space-between/internal';
+import styles from './styles.css.js';
+import { fireNonCancelableEvent } from '../internal/events';
+import { getBaseProps } from '../internal/base-component';
+import checkControlled from '../internal/hooks/check-controlled';
+import clsx from 'clsx';
+import { SomeRequired } from '../internal/types';
+import { warnOnce } from '../internal/logging';
+import { Dropzone, useDropzoneVisible } from './dropzone';
+import FileInput from './file-input';
+import TokenList from '../internal/components/token-list';
+import { Token } from '../token-group/token';
+import { ConstraintText, FormFieldError } from '../form-field/internal';
+import { useUniqueId } from '../internal/hooks/use-unique-id';
+import { useMergeRefs } from '../internal/hooks/use-merge-refs';
+import { joinStrings } from '../internal/utils/strings';
+import { useFormFieldContext } from '../contexts/form-field';
+
+type InternalFileUploadProps = SomeRequired<
+  FileUploadProps,
+  'multiple' | 'showFileSize' | 'showFileLastModified' | 'showFileThumbnail'
+> &
+  InternalBaseComponentProps;
+
+export default React.forwardRef(InternalFileUpload);
+
+function InternalFileUpload(
+  {
+    accept,
+    ariaRequired,
+    multiple,
+    onChange,
+    value,
+    tokenLimit,
+    showFileSize,
+    showFileLastModified,
+    showFileThumbnail,
+    i18nStrings,
+    __internalRootRef = null,
+    constraintText,
+    errorText,
+    fileErrors,
+    ...restProps
+  }: InternalFileUploadProps,
+  externalRef: ForwardedRef<ButtonProps.Ref>
+) {
+  const baseProps = getBaseProps(restProps);
+  const metadata = { showFileSize, showFileLastModified, showFileThumbnail };
+
+  const errorId = useUniqueId('error-');
+  const constraintTextId = useUniqueId('constraint-text-');
+
+  const fileInputRef = useRef<ButtonProps.Ref>(null);
+  const ref = useMergeRefs(fileInputRef, externalRef);
+
+  const [removedFileIndex, setRemovedFileIndex] = useState<null | number>(null);
+  const [removedLastFile, setRemovedLastFile] = useState(false);
+
+  checkControlled('FileUpload', 'value', value, 'onChange', onChange);
+
+  if (!multiple && value.length > 1) {
+    warnOnce('FileUpload', 'Value must be an array of size 0 or 1 when `multiple=false`.');
+  }
+
+  const handleFilesChange = (newFiles: File[]) => {
+    const newValue = multiple ? [...value, ...newFiles] : newFiles[0] ? newFiles : [...value];
+    fireNonCancelableEvent(onChange, { value: newValue });
+  };
+
+  const onFileRemove = (removeFileIndex: number) => {
+    const newValue = value.filter((_, fileIndex) => fileIndex !== removeFileIndex);
+    fireNonCancelableEvent(onChange, { value: newValue });
+    setRemovedFileIndex(removeFileIndex);
+    setRemovedLastFile(value.length === 1);
```

Comment:
This seems like "you might not need an effect" case: https://react.dev/learn/you-might-not-need-an-effect#sharing-logic-between-event-handlers

Is there anything against focusing it right in the handler?

Reply:
P.S. I have questions (simplification options) for `removedFileIndex` too, but since it requires changing current `TokenList` implementation, let's skip this

Reply:
Thanks - I'd love to improve that logic and remove the effects if possible.

The challenge so far is that there must be a render in between remove and focus update. Imagine 5 items are present and 2 are hidden behind "show more". When removing the 3rd item the focus must go to the item which is not yet present in the DOM at that point.

Reply:
`setRemovedLastFile` only needed to move focus from 1 last item to the button. What does it have to do with 5 items use-case?

```suggestion
    if(value.length === 1) {
       // when removing last item, the focus should move to the input
       fileInputRef.current?.focus();
    }
```

Reply:
Ah, this one. Will update that